crates/move-prover-boogie-backend/src/boogie_backend/bytecode_translator.rs

```diff
@@ -2148,12 +2253,255 @@ impl<'env> FunctionTranslator<'env> {
     }
 
     fn should_use_opaque_as_function(&self, write: bool) -> bool {
-        let dinfo: &deterministic_analysis::DeterministicInfo = deterministic_analysis::get_info(self.fun_target.data);
-        let correct_style = self.style == FunctionTranslationStyle::Opaque || (if write { false } else { self.style == FunctionTranslationStyle::SpecNoAbortCheck });
+        let dinfo: &deterministic_analysis::DeterministicInfo =
+            deterministic_analysis::get_info(self.fun_target.data);
+        let correct_style = self.style == FunctionTranslationStyle::Opaque
+            || (if write {
+                false
+            } else {
+                self.style == FunctionTranslationStyle::SpecNoAbortCheck
+            });
 
         dinfo.is_deterministic && correct_style
     }
 
+    /// Check if the function contains any if_then_else operations from conditional merge insertion.
+    fn has_if_then_else_operations(&self) -> bool {
+        use Bytecode::*;
+        self.fun_target.data.code.iter().any(|bytecode| {
+            if let Call(_, _, op, _, _) = bytecode {
+                matches!(op, Operation::IfThenElse)
+            } else {
+                false
+            }
+        })
+    }
+
+    /// Check if the function has early returns (more than one return statement).
+    /// Functions with early returns cannot be emitted as Boogie functions.
+    fn has_early_returns(&self) -> bool {
+        use Bytecode::*;
+        let return_count = self
+            .fun_target
+            .data
+            .code
+            .iter()
+            .filter(|bytecode| matches!(bytecode, Ret(_, _)))
+            .count();
+        return_count > 1
+    }
+
+    /// Check if the function contains any function calls.
+    /// For now, we skip functions with any function calls to avoid transitive requirements.
+    fn has_function_calls(&self) -> bool {
+        use Bytecode::*;
+        self.fun_target.data.code.iter().any(|bytecode| {
+            if let Call(_, _, op, _, _) = bytecode {
+                matches!(op, Operation::Function(_, _, _))
+            } else {
+                false
+            }
+        })
+    }
+
+    /// Gate for emitting an additional Pure function.
+    /// Start conservatively: must contain IfThenElse, have a single return, and no function calls.
+    fn should_emit_pure(&self) -> bool {
```

Comment:
I'm thinking it might be easier for now to use `ext(pure)` attribute on the function, and use that as a trigger instead of very detailed `should_emit_pure`